nengo/builder/optimizer.py

```diff
@@ -0,0 +1,969 @@
+        `ops`.
+        """
+        for op in ops:
+            for key in dir(op):
+                sig = getattr(op, key)
+                if isinstance(sig, Signal):
+                    setattr(op, key, fn(sig))
+
+            op.sets = [fn(s) for s in op.sets]
+            op.incs = [fn(s) for s in op.incs]
+            op.reads = [fn(s) for s in op.reads]
+            op.updates = [fn(s) for s in op.updates]
+
+    @classmethod
+    def _get_sig_view_replacements(cls, ops, sig_replacements):
+        """Given signal replacements `sig_replacements`, returns additional
+        signal replacements required to account for views into a replaced
+        signal.
+
+        Parameters
+        ----------
+        ops : sequence
+            Operators to determine the additional signal replacements for.
+        sig_replacements : dict
+            Signal replacements.
+
+        Returns
+        -------
+        dict
+            Additional required replacements of views into replaced bases of
+            signals.
+        """
+        view_replacements = {}
+
+        def get_sig_view_replacement(s):
+            if s.is_view and s.base in sig_replacements:
+                base_replacement = sig_replacements[s.base]
+                offset = s.offset
+                strides = tuple(a // b * c for a, b, c in zip_longest(
+                    s.strides, s.base.strides, base_replacement.strides,
+                    fillvalue=1))
+                if base_replacement.is_view:
+                    offset += base_replacement.offset
+                    base_replacement = base_replacement.base
+                buf = base_replacement.initial_value
+                initial_value = np.ndarray(
+                    buffer=buf, dtype=s.dtype, shape=s.shape, offset=offset,
+                    strides=strides)
+                view_replacements[s] = Signal(
+                    initial_value, name=s.name, base=base_replacement,
+                    readonly=s.readonly)
+            return s
+
+        cls._map_onto_op_signals(get_sig_view_replacement, ops)
+        return view_replacements
+
+    @classmethod
+    def _replace_op_signals(cls, ops, sig_replacements):
+        """Apply signal repacements `sig_replacements` to all opeators in
+        `ops`.
+        """
+        cls._map_onto_op_signals(lambda s: sig_replacements.get(s, s), ops)
+
+    def _reinitialize_model_ops(self, ops):
+        """Reinitializes the operator list of the builder model from `ops`."""
+        self.model.operators = []
+        for op in ops:
+            self.model.add_op(op)
+
+    def _update_model_sigs(self, sig_replacements):
+        """Updates the signal dictionary in the builder model with
+        `sig_replacements`.
+        """
+        for k in self.model.sig:
+            for name, v in self.model.sig[k].items():
+                self.model.sig[k][name] = sig_replacements.get(v, v)
+
+
+class AbstractMerger(object):
```

Comment:
Right. So maybe move the optimizer to it's own folder so it's components can be split up?